File: sdk/typescript/examples/03_plan_mode.mjs
/**
 * 03-plan-mode — create a session, drive Plan Mode 2.0 (g165–167).
 *
 * Approve or reject the proposed plan with `RecursiveClient.approvePlan` /
 * `rejectPlan`. The server endpoints are:
 *
 *   POST /sessions/{id}/plan/confirm  { edits?: string }
 *   POST /sessions/{id}/plan/reject   { reason?: string }
 *
 * This example assumes the agent will enter `plan_pending_approval` after
 * the first message — adjust the prompt if your config differs.
 */

import { Agent, RecursiveClient } from "../dist/index.mjs";

const baseUrl = process.env.RECURSIVE_BASE_URL ?? "http://127.0.0.1:3000";

const agent = await Agent.create({ baseUrl });
const client = new RecursiveClient({ baseUrl });

try {
  console.log("session :", agent.sessionId);

  const run = await agent.send("Plan: refactor http.rs into smaller modules.");
  // We don't drain the stream here — the server should park in
  // plan_pending_approval as soon as the agent calls exit_plan_mode.
  await run.wait().catch(() => {
    // wait() may reject if the run is paused waiting for plan approval;
    // in that case fall through and inspect session state.
  });

  const detail = await client.getSession(agent.sessionId);
  if (detail.status !== "plan_pending_approval" || !detail.pendingPlan) {
    console.log("session not in plan_pending_approval (status:", detail.status, ")");
    process.exit(0);
  }

  console.log("\n--- pending plan ---");
  console.log(detail.pendingPlan);
  console.log("--------------------\n");

  // Approve as-is. To edit before approving, pass `{ edits: "new plan…" }`.
  const resp = await client.approvePlan(agent.sessionId);
  console.log("approval :", resp.status);

  // Drive the now-approved run to completion.
  const run2 = await agent.send("(continue)");
  await run2.wait();
} finally {
  await agent.close();
}

File: sdk/python/examples/03_plan_mode.py
"""03_plan_mode — drive Plan Mode 2.0 (g165–167) via RecursiveClient.

The agent enters ``plan_pending_approval`` after calling ``exit_plan_mode``.
Approve or reject via these endpoints:

    POST /sessions/{id}/plan/confirm  { edits?: str }
    POST /sessions/{id}/plan/reject   { reason?: str }
"""

from __future__ import annotations

import os
import sys

from recursive_client import RecursiveClient
from recursive_sdk import Agent


def main() -> int:
    base_url = os.environ.get("RECURSIVE_BASE_URL", "http://127.0.0.1:3000")
    client = RecursiveClient(base_url=base_url)

    with Agent.create(base_url=base_url) as agent:
        print("session :", agent.session_id)

        run = agent.send("Plan: refactor http.rs into smaller modules.")
        # The run may park in plan_pending_approval before completing —
        # swallow any wait() failure and inspect the session state directly.
        try:
            run.wait()
        except Exception:
            pass

        detail = client.get_session(agent.session_id)
        if detail.status != "plan_pending_approval" or not detail.pending_plan:
            print("session not in plan_pending_approval (status:", detail.status, ")")
            return 0

        print("\n--- pending plan ---")
        print(detail.pending_plan)
        print("--------------------\n")

        # Approve as-is. To edit before approving, pass edits=...
        resp = client.approve_plan(agent.session_id)
        print("approval :", resp["status"])

        # Drive the now-approved run to completion.
        agent.send("(continue)").wait()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
